Show scancodes in debugfs

dvad924 opened the pull request:
Log only codes for flexibility #3 

I added shift masks as well figure it might make parsing combos easier.

- Feature Description
     Add option to log the raw keycodes instead of the US keyboard characters

- Design Changes
     An optional integer parameter can be passed at time of inserting the module 
     to log raw codes. Ex : `sudo insmod kisni.ko codes=1` If the parameter is not included
     the module will default to logging US characters. Using this option will result in the log file
     having each line formatted as "shiftmask_space_keycode" where shiftmask and keycode 
     are the base 10 representational strings of the values received. The shiftmask was added for 
     use in finding key combinations.

     Integer module parameter added.
     String values are written to the same buffer in the same way as the original implementation.

dvad924 committed:
Changed US keys to keycode, also added shift mask

dvad924 committed:
added the module param

jarun commented on keysniffer.c:
```diff
@@ -102,12 +107,45 @@ static struct notifier_block keysniffer_blk = {
+  int remainders[size];
+  int start;
+  /* zero out the buffer */
+  for ( i = 0; i < size; i++)
```
Extra space between `( and i`.

jarun commented on keysniffer.c:
```diff
@@ -102,12 +107,45 @@ static struct notifier_block keysniffer_blk = {
+  for ( i = 0; i < size; i++)
+    rep[i] = 0;
+
+  if( !input ){
```
There should be a space between `if and (`

jarun commented:
Please follow the kernel coding guideline:

https://github.com/torvalds/linux/blob/master/Documentation/process/coding-style.rst

The basics should cover. If that seem time-consuming, just refer to a random file, e.g.

https://github.com/torvalds/linux/blob/master/crypto/ablk_helper.c

dvad924 commented:
Thanks for the reference!
I will get on that, should be fixed soon.

dvad924 committed:
Tweaked style to fit standard

jarun commented on keysniffer.c:
```diff
@@ -107,95 +108,102 @@ static struct notifier_block keysniffer_blk = {
+ *  - rep   : string buffer to place the representation in
+ *  - size  : the max size of the buffer
+ */
+void to_base_n_string(int input, int n,char *rep,int size){
```
Please add spaces between `int n,char *rep,int size)`
The `{` should be on the next line.

jarun commented on keysniffer.c:
```diff
@@ -107,95 +108,102 @@ static struct notifier_block keysniffer_blk = {
+
+	i = size - 1;
+
+	while ( div > 0 ) {
```
No extra spaces after `(` and before `)`.

jarun commented on keysniffer.c:
```diff
@@ -107,95 +108,102 @@ static struct notifier_block keysniffer_blk = {
+
+	start = i+1;
+
+	for (i = start; i < size; i++) {
```
You can drop `{` and `}` for single liners.

jarun commented on keysniffer.c:
```diff
@@ -107,95 +108,102 @@ static struct notifier_block keysniffer_blk = {
+			: us_keymap[param->value][0];
+		if (pressed_key) {
+			int pklen, replen;
+			if ( codes ) {
```
No extra spaces after `(` and before `)`.

jarun commented on keysniffer.c:
```diff
@@ -107,95 +108,102 @@ static struct notifier_block keysniffer_blk = {
+				buf_pos = 0;
+			}
+
+			if( codes ) {
```
Space after `if`.

jarun commented on keysniffer.c:
```diff
@@ -102,35 +108,98 @@ static struct notifier_block keysniffer_blk = {
 		if (pressed_key) {
-			len = strlen(pressed_key);
+			int pklen, replen;
+			if ( codes ) {
```
No extra spaces after `(` and before `)`.

jarun commented on keysniffer.c:
```diff
@@ -102,35 +108,98 @@ static struct notifier_block keysniffer_blk = {
+				replen = strlen(rep);
+
+				len = pklen + replen + 1;
+			} else {
```
You can drop `{` and `}` for single liners.

jarun commented on keysniffer.c:
```diff
@@ -102,35 +108,98 @@ static struct notifier_block keysniffer_blk = {
-			len = strlen(pressed_key);
+			int pklen, replen;
+			if ( codes ) {
+				to_base_n_string(param->value,10,key,sizeof(int)+1);
```
`(param->value,10,key,sizeof(int)+1)`

should be

`(param->value, 10, key, sizeof(int) + 1)`

jarun commented on keysniffer.c:
```diff
@@ -102,35 +108,98 @@ static struct notifier_block keysniffer_blk = {
+			int pklen, replen;
+			if ( codes ) {
+				to_base_n_string(param->value,10,key,sizeof(int)+1);
+				to_base_n_string(param->shift,10,rep,sizeof(int)+1);
```
`(param->shift,10,rep,sizeof(int)+1)`

should be

`(param->shift, 10, rep, sizeof(int) + 1)`

jarun commented on keysniffer.c:
```diff
@@ -102,35 +108,98 @@ static struct notifier_block keysniffer_blk = {
+				buf_pos += replen;
+				strncpy(keys_buf + buf_pos, " ", 1);
+				buf_pos++;
+				strncpy(keys_buf + buf_pos, pressed_key,
```
It's OK to bring the last param in the same line instead of the next line.

jarun commented:
I have reviewed as much as possible. But more violations are possible given you are apparently writing kernel module code for the first time. Please use [checkpatch](http://tuxdiary.com/2015/03/22/check-kernel-code-checkpatch/) to fix the remaining defects. Before accepting the PR, I'll verify against it too.

dvad924 committed:
Used removed all code formatting errors via checkpatch.pl

jarun commented on keysniffer.c:
```diff
@@ -133,17 +134,17 @@ void to_base_n_string(int input, int n,char *rep,int size){
+	for (i = start; i < size; i++)
 		rep[i - start] = '0' + remainders[i];
-	}
+
```
Redundant blank line?

jarun commented on keysniffer.c:
```diff
@@ -102,35 +108,99 @@ static struct notifier_block keysniffer_blk = {
+
+	for (i = start; i < size; i++)
+		rep[i - start] = '0' + remainders[i];
+
```
Redundant newline.

jarun commented on keysniffer.c:
```diff
@@ -102,35 +108,99 @@ static struct notifier_block keysniffer_blk = {
+				buf_pos++;
+				strncpy(keys_buf + buf_pos, pressed_key, pklen);
+				buf_pos += pklen;
+
```
extra blank line

dvad924 commented:
Thanks for all the feedback!
Yes this is my first module.
I apologize for not adhering strictly before, I thought I hit the major points but didn't realize how important the formatting is to the code. I used the checkpatch.pl script and removed all the errors.
(there are still warnings for lines over 80chars and not using the "const" keyword)

jarun commented:
I also need a **detailed explanation** of what you are trying to do in the code, usually something like an elaborate commit message. For example, how you are encoding the keypresses, how you are decoding them when read from debugfs etc.

I would appreciate if you realize the fact that PRs without any description of how it is achieving something seems meaningless to people who would be reviewing them.

dvad924 committed:
Add feature log raw keycodes

Add a module parameter to set if the module
will log the US Keyboard characters
or the raw keycode values. In raw mode the
shift mask is also added to beginning of each line.

This is done by adding the integer flag "codes" as
a module parameter. If the flag is 0 (default) the behavior
is unchanged from the original version. Otherwise
the shift mask and keycode are converted to their
base 10 string representations and written to the
"keys" log file in the format "mask_space_keycode".

jarun commented:
I wasn't referring to cod comments... A commit description as a whole. For example, take a look at this PR: https://github.com/jarun/Buku/pull/128

Here the contributor has explained in complete detail what he is trying to do and how he is doing it. I need something at this level to *start* reviewing the logic in the code.

jarun commented on README.md:
```diff
@@ -52,6 +52,33 @@ To view the pressed keys, run:
+    1 33
+    0 14
+    0 29
+    4 30
```
What is 4? As per my understanding, this is shift_mask and can be 0 or 1 right?

dvad924 replied:
Originally I thought so as well, but per testing it seems to capture Shift,Ctrl,Alt possibly more, haven't found docs on it yet, keyboard.h doesn't have any insights. It is a proper bitmask it seems.

jarun replied:
Strange indeed. Thanks! We should show these values.

jarun commented on EX.txt:
```diff
@@ -0,0 +1,10 @@
+1 52
```
What is this new file EX.txt?

dvad924 replied:
This was an example I included before adding to the readme, I will remove it.

jarun replied:
Thanks!

jarun commented on keysniffer.c:
```diff
@@ -104,33 +110,56 @@ static struct notifier_block keysniffer_blk = {
 	if (!(param->down))
 		return NOTIFY_OK;
 
 	if (param->value >= 0x1 && param->value <= 0x77) {
```
If you want to show any value, why do you put your logic within this condition?

dvad924 replied:
This was done with the intent of restricting the characters logged to the same as was done originally. If we want any & all codes that is great and should be done in a more modular function that could deal with a number of different parameters and options

jarun replied:
Yes, we need to keep it distinct. If it's possible to fetch the max and min scancodes from the keyboard driver, our life would be easy to limit the buffer size required to hold a scancode and the boundaries for our check.

jarun commented on keysniffer.c:
```diff
@@ -104,33 +110,56 @@ static struct notifier_block keysniffer_blk = {
+		  void *_param)
 {
 	size_t len;
+	int max_len = sizeof(int) + 1;
```
Will it be possible to capture Unicode? Also can we show the values in hex rather than in decimal? We can omit the 0x prefix in all cases.

dvad924 replied:
Hex is fine, Unicode I will need to look into ...

jarun replied:
Sure, it would be a great value addition.

dvad924 commented:
- Feature Description
     Add option to log the raw keycodes instead of the US keyboard characters

- Design Changes
     An optional integer parameter can be passed at time of inserting the module 
     to log raw codes. Ex : `sudo insmod kisni.ko codes=1` If the parameter is not included
     the module will default to logging US characters. Using this option will result in the log file
     having each line formatted as "shiftmask_space_keycode" where shiftmask and keycode 
     are the base 10 representational strings of the values received. The shiftmask was added for 
     use in finding key combinations.

     Integer module parameter added.
     String values are written to the same buffer in the same way as the original implementation.

jarun commented:
Note: If it's possible to fetch the max and min scancode from the keyboard driver, our life would be easy to limit the buffer size required to hold a scancode.

jarun commented:
How about the following (though it does not break out of the 0x1 to 0x77 limit yet):

```
diff --git a/keysniffer.c b/keysniffer.c
index 0fec77a..4138d15 100644
--- a/keysniffer.c
+++ b/keysniffer.c
@@ -21,16 +21,23 @@
 #include <linux/init.h>
 #include <linux/kernel.h>
 #include <linux/module.h>
+#include <linux/moduleparam.h>
 #include <linux/keyboard.h>
 #include <linux/debugfs.h>
 
 #define BUF_LEN (PAGE_SIZE << 2) /* 16KB buffer (assuming 4KB PAGE_SIZE) */
 
+#define MAX_ENC_LEN 12           /* Length of an encoded 'code shift' chunk */
+static int codes;                /* Log type module parameter */
+
 MODULE_LICENSE("GPL v2");
 MODULE_AUTHOR("Arun Prakash Jana <engineerarun@gmail.com>");
 MODULE_VERSION("1.4");
 MODULE_DESCRIPTION("Sniff and log keys pressed in the system to debugfs");
 
+module_param(codes, int, 0644);
+MODULE_PARM_DESC(codes, "Flag: 0: US keyboard, 1: keycodes (base10)");
+
 /* Declarations */
 static struct dentry *file;
 static struct dentry *subdir;
@@ -104,24 +111,34 @@ static struct notifier_block keysniffer_blk = {
 
 /* Keypress callback */
 int keysniffer_cb(struct notifier_block *nblock,
-		unsigned long code,
-		void *_param)
+		  unsigned long code,
+		  void *_param)
 {
 	size_t len;
+	char code_buf[MAX_ENC_LEN] = {0};
 	struct keyboard_notifier_param *param = _param;
 	const char *pressed_key;
 
 	pr_debug("code: 0x%lx, down: 0x%x, shift: 0x%x, value: 0x%x\n",
-		code, param->down, param->shift, param->value);
+		 code, param->down, param->shift, param->value);
 
 	if (!(param->down))
 		return NOTIFY_OK;
 
 	if (param->value >= 0x1 && param->value <= 0x77) {
 		pressed_key = param->shift
-				? us_keymap[param->value][1]
-				: us_keymap[param->value][0];
+			? us_keymap[param->value][1]
+			: us_keymap[param->value][0];
 		if (pressed_key) {
+			if (codes) {
+				sprintf(code_buf, "%x ", param->value);
+                                if (param->shift)
+                                    strncat(code_buf, "1", MAX_ENC_LEN);
+                                else
+                                    strncat(code_buf, "0", MAX_ENC_LEN);
+				pressed_key = code_buf;
+			}
+
 			len = strlen(pressed_key);
 
 			if ((buf_pos + len) >= BUF_LEN) {
```

dvad924 committed:
Removed unecessary example

dvad924 committed:
Add fun c -> str (still needs code format)

Was on a different machine, will reformat when I have the check script

jarun commented on keysniffer.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * A Linux kernel module to grab keycodes and log to debugfs
+1;2802;0c * A Linux kernel module to grab keycodes and log to debugfs
```
What's this? ;)

jarun commented on keysniffer.c:
```diff
@@ -21,16 +21,31 @@
+#define DEC 2                         /* Type code for decimal log */
+#define OCT 3                         /* Type code for octal log */
+#define UNI 4                         /* Type code for Unicode log */
+#define SHIFT 1                       /* shift_mask code for shift */
```
We are not using this. Please remove.

jarun commented on keysniffer.c:
```diff
@@ -102,40 +117,69 @@ static struct notifier_block keysniffer_blk = {
+  case DEC:
+    format_type = "%d %d";
+    break;
+  case OCT:
```
We don't need octal. Please remove.

jarun commented on keysniffer.c:
```diff
@@ -102,40 +117,69 @@ static struct notifier_block keysniffer_blk = {
+  case OCT:
+    format_type = "%o %o";
+    break;
+  case UNI:
```
As of now it is same as HEX and probably we need to handle Unicode differently. Please remove.

jarun commented on keysniffer.c:
```diff
@@ -21,16 +21,31 @@
 MODULE_DESCRIPTION("Sniff and log keys pressed in the system to debugfs");
 
+module_param(codes, int, 0644);
+MODULE_PARM_DESC(codes, "Flag: 0: US keyboard, 1: keycodes (base16) 2: (base10) 3: (base8)");
```
Please mention 0 as default, i.e.:

`Flag: 0: US keyboard (default)`

Also, please drop octal. I believe the desc could be:

`log format, 0 = US keyboard (default), 1 = hex keycodes, 2 = dec keyodes (int)`

jarun commented on keysniffer.c:
```diff
@@ -102,40 +117,69 @@ static struct notifier_block keysniffer_blk = {
+			? us_keymap[keycode][1]
+			: us_keymap[keycode][0];
+      
+      snprintf(buffer, MAX_ENC_LEN, "%s %x", pressed_key, shift_mask);
```
Nice! I forgot snprintf() can handle multiple args! :+1:

jarun commented:
The last commit is very close and quite clean. We're almost there. :+1:

jarun commented:
It's been a long time. I'll take care of the comments myself.

jarun commented:
The indentation, coding style (tabs converted into spaces or just new spaces) is broken in every function this PR touches. Reverting.

dvad924 commented:
Thanks for finishing it and sorry I wasn't able to get to in the past few days, had other things going on.

jarun commented:
No problem!
